### DISCUSSION TITLE
How/when are capsules disposed?

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#31] [Q&A] [answered]
```dart
/// Registers the given [SideEffectApiCallback]
/// to be called on capsule disposal.
void registerDispose(SideEffectApiCallback callback);
```

When/how a capsule is disposed?
Only when called `CapsuleContainer.dispose()`?
How to dispose a container in Flutter?

**Comment by GregoryConrad** [comment, marked as answer]
﻿It depends on the capsule.

- idempotent (does not register side effects): may be disposed at any time when there is no “demand” (in practice, they are disposed when an upstream capsule returns new data)
- non-idempotent (registers side effect(s)): will live from the time they are first read until the container is disposed

in flutter, RearchBootstrapper will manage the container life cycle for you. You should not manually dispose a container in flutter
